src/liballoc/collections/btree/map.rs

```diff
@@ -2044,65 +2044,51 @@ where
-    let mut diverged = false;
 
     loop {
-        let min_edge = match (min_found, range.start_bound()) {
-            (false, Included(key)) => match search::search_linear(&min_node, key) {
-                (i, true) => {
+        let front = match (min_found, range.start_bound()) {
+            (false, Included(key)) => match search::search_node(min_node, key) {
+                Found(kv) => {
                     min_found = true;
-                    i
+                    kv.left_edge()
                 }
-                (i, false) => i,
+                GoDown(edge) => edge,
             },
-            (false, Excluded(key)) => match search::search_linear(&min_node, key) {
-                (i, true) => {
+            (false, Excluded(key)) => match search::search_node(min_node, key) {
+                Found(kv) => {
                     min_found = true;
-                    i + 1
+                    kv.right_edge()
                 }
-                (i, false) => i,
+                GoDown(edge) => edge,
             },
-            (_, Unbounded) => 0,
-            (true, Included(_)) => min_node.len(),
-            (true, Excluded(_)) => 0,
+            (true, Included(_)) => min_node.last_edge(),
+            (true, Excluded(_)) => min_node.first_edge(),
+            (_, Unbounded) => min_node.first_edge(),
         };
 
-        let max_edge = match (max_found, range.end_bound()) {
-            (false, Included(key)) => match search::search_linear(&max_node, key) {
-                (i, true) => {
+        let back = match (max_found, range.end_bound()) {
+            (false, Included(key)) => match search::search_node(max_node, key) {
+                Found(kv) => {
                     max_found = true;
-                    i + 1
+                    kv.right_edge()
                 }
-                (i, false) => i,
+                GoDown(edge) => edge,
             },
-            (false, Excluded(key)) => match search::search_linear(&max_node, key) {
-                (i, true) => {
+            (false, Excluded(key)) => match search::search_node(max_node, key) {
+                Found(kv) => {
                     max_found = true;
-                    i
+                    kv.left_edge()
                 }
-                (i, false) => i,
+                GoDown(edge) => edge,
             },
-            (_, Unbounded) => max_node.len(),
-            (true, Included(_)) => 0,
-            (true, Excluded(_)) => max_node.len(),
+            (true, Included(_)) => max_node.first_edge(),
+            (true, Excluded(_)) => max_node.last_edge(),
+            (_, Unbounded) => max_node.last_edge(),
         };
 
-        if !diverged {
-            if max_edge < min_edge {
-                panic!("Ord is ill-defined in BTreeMap range")
-            }
-            if min_edge != max_edge {
-                diverged = true;
-            }
+        if front.partial_cmp(&back) == Some(Ordering::Greater) {
```

Comment:
AFAICT, this is now checking more often. Presumably, the old code was not checking as an optimization, since we can (I believe) see that if the edges are different, and since this is a tree, we are guaranteed to not cross over again.

I guess based on benchmark results that doesn't matter in practice (unsurprisingly, to me), so up to you whether we leave the old behavior or switch to the more stringent checking that is currently in this PR.

Reply:
Well, checking more... it's comparing node pointers (and nothing else once it's diverged) instead of checking a boolean. I think the node pointers should be hot in the cache, and by not having the boolean there's a little more room in that cache. As it happens, I wondered the same thing and put back in the `diverged` flag, also to confirm that it does the same thing. I didn't make a difference. I just rebased that branch and this PR, compared again, and both have the same performance as master.

I doubt it was an optimization. I think the author didn't want/dare to implement an equality predicate for nodes (there's one for handles, not for NodeRef), and comparing edge indices in the previous iteration was very easy. I just checked file history and it was first committed like that, so we'll never know.

Reply:
Makes sense. This is fine to leave in as I mentioned. Thanks for taking a look!